baserun/instrumentation/instrumented_wrapper.py

```diff
@@ -0,0 +1,120 @@
+            return wrapped_fn(*args, **kwargs)
+
+        parent_span: _Span = get_current_span()
+        request_type = span_name.split(".")[-1]
+        parent_span.set_attribute(SpanAttributes.LLM_REQUEST_TYPE, request_type)
+
+        # Start new span
+        span = tracer.start_span(
+            f"baserun.{span_name}",
+            kind=SpanKind.CLIENT,
+            attributes=parent_span.attributes,
+        )
+
+        auto_end_span = True
+        try:
+            # Activate the span in the current context, but don't end it automatically
+            with trace.use_span(span, end_on_exit=False):
+                # Capture request attributes
+                # noinspection PyBroadException
+                try:
+                    instrumentor.set_request_attributes(span, kwargs)
+
+                    span.set_attribute(
+                        SpanAttributes.LLM_REQUEST_MODEL, kwargs.get("model")
+                    )
+
+                    max_tokens = kwargs.get(
+                        "max_tokens", kwargs.get("max_tokens_to_sample")
+                    )
+                    if max_tokens is not None:
+                        span.set_attribute(
+                            SpanAttributes.LLM_REQUEST_MAX_TOKENS, max_tokens
+                        )
+                except Exception as e:
+                    logger.warning(
+                        f"Failed to set input attributes for Baserun span, error: {e}"
+                    )
+
+                # Actually call the wrapped method
+                response = wrapped_fn(*args, **kwargs)
+
+                # If this is a streaming response, wrap it, so we can capture each chunk
+                if isinstance(response, GeneratorType):
+                    wrapped_response = instrumentor.generator_wrapper(response, span)
+                    # The span will be ended inside the generator once it's finished
+                    auto_end_span = False
+                    return wrapped_response
+
+                run = Baserun.current_run()
+                if not run:
+                    logger.warning(
+                        "Baserun data not propagated correctly, cannot send data"
+                    )
+                    return response
+
+                span.set_attribute(
+                    SpanAttributes.BASERUN_RUN, Baserun.serialize_run(run)
+                )
+
+                # If it's a full response capture the response attributes
+                if response:
+                    # noinspection PyBroadException
+                    try:
+                        span.set_status(Status(StatusCode.OK))
+                        instrumentor.set_response_attributes(span, response)
+                    except Exception as e:
+                        logger.warning(
+                            f"Failed to set response attributes for Baserun span, error: {e}"
+                        )
+                else:
```

Comment:
I don't think this is possible? The wrapped_fn would have thrown an exception and that should have bubbled up to our exception handling in baserun.py?

Reply:
This is mostly if the user didn't return anything from their traced function. Just a way to clarify why we won't have response attributes when they're first doing setup.